File: leetcode/validParentheses.js
/*
Valid Parentheses
https://leetcode.com/problems/valid-parentheses/
Optimize!
*/
var isValid = function (s) {
  let stack = [];
  for (let i = 0; i < s.length; i++) {
    if (stack[stack.length - 1] === "(" && s[i] === ")") {
      stack.pop();
    } else if (stack[stack.length - 1] === "{" && s[i] === "}") {
      stack.pop();
    } else if (stack[stack.length - 1] === "[" && s[i] === "]") {
      stack.pop();
    } else {
      stack.push(s[i]);
    }
  }
  return stack.length === 0;
};

console.log(isValid("(){}[]")); // true
console.log(isValid("]")); // false

File: neetcode/stacks/valid_parentheses.py
'''
Valid Parentheses
https://leetcode.com/problems/valid-parentheses
https://neetcode.io/problems/validate-parentheses
'''
def isValid(s) -> bool:
    stack = []
    pairs = {")": "(", "}": "{", "]": "["}
    for c in s:
        if c in pairs:
            if stack and stack[-1] == pairs[c]:
                stack.pop()
            else:
                return False
        else:
            stack.append(c)
    return len(stack) == 0

print(isValid("()[]{}")) # True
print(isValid("(]")) # False